packages/from-hardhat/src/compilation.ts

```diff
@@ -0,0 +1,121 @@
+import type * as Hardhat from "hardhat/types";
+
+import type { Compilation, CompiledContract } from "@truffle/compile-common";
+import * as CompileSolidity from "@truffle/compile-solidity";
+
+import { supportedHardhatBuildInfoFormats } from "./constants";
+
+export const fromBuildInfo = (buildInfo: Hardhat.BuildInfo): Compilation => {
+  const { _format } = buildInfo;
+
+  if (!supportedHardhatBuildInfoFormats.has(_format)) {
+    throw new Error(`Unsupported build info format: ${_format}`);
+  }
+
+  const sourceIndexes = SourceIndexes.fromBuildInfo(buildInfo);
+
+  return {
+    sourceIndexes,
+    sources: Sources.fromBuildInfo(buildInfo, sourceIndexes),
+    contracts: Contracts.fromBuildInfo(buildInfo),
+    compiler: {
+      name: "solc",
+      version: buildInfo.solcLongVersion
+    }
+  };
+};
+
+namespace SourceIndexes {
+  export const fromBuildInfo = (
+    buildInfo: Hardhat.BuildInfo
+  ): Compilation["sourceIndexes"] => {
+    const sourceIndexes = [];
+    for (const { index, sourcePath } of Object.entries(
+      buildInfo.output.sources
+    ).map(([sourcePath, source]) => ({ index: source.id, sourcePath }))) {
+      sourceIndexes[index] = sourcePath;
+    }
+
+    return sourceIndexes;
+  };
+}
+
+namespace Sources {
+  export const fromBuildInfo = (
+    buildInfo: Hardhat.BuildInfo,
+    sourceIndexes: Compilation["sourceIndexes"]
+  ): Compilation["sources"] =>
+    sourceIndexes.map(sourcePath => {
+      // to handle if sourceIndexes is a sparse array
+      if (!sourcePath) {
+        return;
+      }
+
+      const inputSource = buildInfo.input.sources[sourcePath];
+      const outputSource = buildInfo.output.sources[sourcePath];
+
+      return {
+        sourcePath,
+        contents: inputSource.content,
+        ast: outputSource.ast,
+        language: buildInfo.input.language
+      };
+    });
+}
+
+namespace Contracts {
+  export const fromBuildInfo = (
+    buildInfo: Hardhat.BuildInfo
+  ): CompiledContract[] => {
+    const contracts = [];
+    for (const [sourcePath, sourceContracts] of Object.entries(
+      buildInfo.output.contracts
+    )) {
+      for (const [contractName, compilerOutputContract] of Object.entries(
+        sourceContracts
+      )) {
```

Comment:
Are we not worried about sparse arrays here?

Also if you want to remain consistent with the functional style above, you could do something like 

```typescript
Object.entries(buildInfo.output.contracts).map([sourcePath, sourceContracts] => 
  Object.entries(sourceContracts).map([contractName, compilerOutputContract] => return { 
    /* contents of CompiledContract here */
  }).flat();
```

Reply:
Though that assumes that we aren't working with sparse arrays.

Reply:
I don't think it's at all necessary to "fix" this, as it's not really an issue. Feel free to resolve immediately if you don't wanna fix it, or resolve after pushing if you do.

Reply:
Resolving :)